### PR TITLE
IGNITE-21245 Remove APPLIED_REV_KEY from metastorage manager

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-21245

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java [hunk 1]
```diff
@@ -121,12 +105,17 @@ public class MetaStorageManagerImpl implements MetaStorageManager {
...
+    private final AtomicLong appliedRevision = new AtomicLong(0L);
+
     /**
      * Future which completes when MetaStorage manager finished local recovery. The value of the future is the revision which must be used
      * for state recovery by other components.
      */
     private final CompletableFuture<Long> recoveryFinishedFuture = new CompletableFuture<>();
 
+    private final CompletableFuture<Long> recoveryFinishedPublicFuture
+            = recoveryFinishedFuture.whenComplete((revision, e) -> appliedRevision.set(revision));
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:117]: maybe you intended to use `thenAccept` here?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java [hunk 2]
```diff
@@ -121,12 +105,17 @@ public class MetaStorageManagerImpl implements MetaStorageManager {
     /** Prevents double stopping of the component. */
     private final AtomicBoolean isStopped = new AtomicBoolean();
 
+    private final AtomicLong appliedRevision = new AtomicLong(0L);
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:108]: what was wrong with `volatile long`?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java [hunk 3]
```diff
@@ -105,7 +106,8 @@ public interface EntryReader {
     public WatchProcessor(String nodeName, EntryReader entryReader) {
         this.entryReader = entryReader;
 
-        this.watchExecutor = Executors.newFixedThreadPool(4, NamedThreadFactory.create(nodeName, "metastorage-watch-executor", LOG));
+        this.watchExecutor = Executors.newSingleThreadScheduledExecutor(
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java:109]: What is this change for?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java [hunk 4]
```diff
@@ -284,16 +288,10 @@ private CompletableFuture<List<WatchAndEvents>> collectWatchesAndEvents(List<Ent
...
         try {
-            // Only notify about entries that have been accepted by at least one Watch.
-            var acceptedEntries = new HashSet<EntryEvent>();
-
-            for (WatchAndEvents watchAndEvents : watchAndEventsList) {
-                acceptedEntries.addAll(watchAndEvents.events);
-            }
-
-            var event = new WatchEvent(acceptedEntries, revision, time);
+            // We consciously put empty set here. Revision applied callback doesn't need any data.
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java:293]: I would suggest to change the `revisionCallback` interface explicitly only require the revision and the timestamp

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/WatchProcessorTest.java [hunk 5]
```diff
@@ -98,7 +99,7 @@ void testGroupEventNotification() {
 
         WatchEvent event = watchEventCaptor.getValue();
 
-        assertThat(event.entryEvents(), containsInAnyOrder(entryEvent1, entryEvent2));
+        assertThat(event.entryEvents(), is(equalTo(emptyList())));
```
Comment (sashapolo) [inline modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/WatchProcessorTest.java:102]: I believe there should be Matcher that looks something like `is(empty())`

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java [hunk 6]
```diff
@@ -449,7 +422,7 @@ public CompletableFuture<Void> deployWatches() {
         }
 
         try {
-            return recoveryFinishedFuture
+            return recoveryFinishedPublicFuture
                     .thenAccept(revision -> inBusyLock(busyLock, () -> {
                         // Meta Storage contract states that all updated entries under a particular revision must be stored in the Vault.
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:427]: Also, this comment is obsolete

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java (hunk 1 above)
Comment (ibessonov) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:117]: Maybe

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java (hunk 2 above)
Comment (ibessonov) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:108]: Mirza knows. I'll add a comment

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java (hunk 3 above)
Comment (ibessonov) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java:109]: Slipped accidentally. Anyway, only 1 thread is used in reality, others do nothing, you can clearly see that in a profiler

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java (hunk 4 above)
Comment (ibessonov) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java:293]: Ok, I'll do it

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java (hunk 1 above)
Comment (sashapolo) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:117]: Can `recoveryFinishedFuture` be completed with a exception? If yes, then the current approach is incorrect (you should verify that `e != null`)

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java (hunk 3 above)
Comment (sashapolo) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java:108]: yeah, the idea was to be able to process multiple watches for a single event simultaneously

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java (hunk 2 above)
Comment (sashapolo) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:108]: > primitive fields won't allow using {@code Mockito#spy(metastorage)}

What do you mean?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java [hunk 7]
```diff
@@ -174,7 +173,7 @@ public CompletableFuture<Void> notifyWatches(List<Entry> updatedEntries, HybridT
                                 CompletableFuture<Void> notifyUpdateRevisionFuture = notifyUpdateRevisionListeners(newRevision);
 
                                 CompletableFuture<Void> notificationFuture = allOf(notifyWatchesFuture, notifyUpdateRevisionFuture)
-                                        .thenComposeAsync(
+                                        .thenAcceptAsync(
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java:176]: you can use `thenRunAsync` instead

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java [hunk 8]
```diff
@@ -288,23 +287,15 @@ private CompletableFuture<List<WatchAndEvents>> collectWatchesAndEvents(List<Ent
...
 
-            return revisionCallback.onRevisionApplied(event)
-                    .whenComplete((ignored, e) -> {
-                        if (e != null) {
-                            LOG.error("Error occurred when notifying watches", e);
-                        }
-                    });
+            revisionCallback.onRevisionApplied(revision);
         } catch (Throwable e) {
             LOG.error("Error occurred when notifying watches", e);
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java:296]: I think we can remove this catch and use `whenComplete` above to log any errors

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java (hunk 3 above)
Comment (sashapolo) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java:108]: If you are going to leave this change, then the whole class should be refactored a bit

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java [hunk 9]
```diff
@@ -173,12 +172,20 @@ public CompletableFuture<Void> notifyWatches(List<Entry> updatedEntries, HybridT
...
                                         );
 
-                                notificationFuture.whenComplete((unused, e) -> maybeLogLongProcessing(updatedEntries, startTimeNanos));
+                                notificationFuture.whenComplete((unused, e) -> {
+                                    maybeLogLongProcessing(updatedEntries, startTimeNanos);
+
+                                    if (e != null) {
+                                        LOG.error("Error occurred when notifying watches", e);
+
+                                        throw new CompletionException(e);
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java:186]: This `throw` is redundant, it won't be propagated anywhere

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java [hunk 10]
```diff
@@ -855,19 +824,19 @@ private void onSafeTimeAdvanced(HybridTimestamp time) {
     }
 
     /**
-     * Saves processed Meta Storage revision and corresponding entries to the Vault.
+     * Saves processed Meta Storage revision to the {@link #appliedRevision}.
      */
-    private CompletableFuture<Void> onRevisionApplied(WatchEvent watchEvent) {
+    private CompletableFuture<Void> onRevisionApplied(long revision) {
```
Comment (sashapolo) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java:829]: 1. This method doesn't need to return a future anymore.
2. All it does is assigning a field, I guess we no longer need the busy lock as well.

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java (hunk 9 above)
Comment (ibessonov) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/WatchProcessor.java:186]: That's right, I feel stupid now

**Review by sashapolo — APPROVED** [review]

**Pull request merged**